### ISSUE TITLE
Wrong result of `table.indexof()`

### ISSUE THREAD

**Issue by alek13** [luanti-org/luanti#14906]
### Minetest version

```true
Minetest 5.9.0-dev-9a1501ae8
```

### Summary

```lua
	print(dump(table.indexof({
		[2] = "lottores:salt",
		[4] = "group:mushroom",
		[5] = "group:mushroom",
		[6] = "group:mushroom",
		[8] = "lottfarming:bowl"
	}, "lottores:salt")))
```
output:
```
-1
```
expected:
```
2
```

**Comment by Emojigit** [comment]
This is intentional, or at least that's Lua's fault. The documentation says:

> (`ipairs`) ... will iterate over the pairs (`1,t[1]`), (`2,t[2]`), ···, up to the **first integer key absent** from the table.

Because your table lacks the first integer key, the iteration stops without working on any values. As Lua arrays are not meant to have holes, IMO this issue is invalid and thus can be closed.

**Comment by rubenwardy** [comment]
This is international, indexof only works with array-like tables. The table you specified is a dictionary/hash-like table due to the gaps. The documentation could be improved to say this

**Comment by appgurueu** [comment]
> If we decide to let it work on dictionaries (or "arrays" with holes) too, replace the `ipairs` with `pairs` and it would work.

Not quite. We guarantee that we find the smallest numerical index. So if we were to support this, we'd have to take the min of all indices with matching values, since `pairs` guarantees no particular order of iteration.

**Comment by rubenwardy** [comment]
That would also make sense, could have no guarantee on ordering

**Comment by appgurueu** [comment]
`table.keyof` happened, list is well-defined.
